samza-kafka/src/main/scala/org/apache/samza/system/kafka/KafkaConsumerProxy.java

```diff
@@ -0,0 +1,444 @@
+    long highWatermark = recordOffset + currentSSPLag; // derived value for the highwatermark
+
+    int size = getRecordSize(r);
+    kafkaConsumerMetrics.incReads(tap);
+    kafkaConsumerMetrics.incBytesReads(tap, size);
+    kafkaConsumerMetrics.setOffsets(tap, recordOffset);
+    kafkaConsumerMetrics.incClientBytesReads(metricName, size);
+    kafkaConsumerMetrics.setHighWatermarkValue(tap, highWatermark);
+  }
+
+
+  private void moveMessagesToTheirQueue(SystemStreamPartition ssp, List<IncomingMessageEnvelope> envelopes) {
+    long nextOffset = nextOffsets.get(ssp);
+
+    for (IncomingMessageEnvelope env : envelopes) {
+      sink.addMessage(ssp, env);  // move message to the BlockingEnvelopeMap's queue
+
+      LOG.trace("IncomingMessageEnvelope. got envelope with offset:{} for ssp={}", env.getOffset(), ssp);
+      nextOffset = Long.valueOf(env.getOffset()) + 1;
+    }
+
+    nextOffsets.put(ssp, nextOffset);
+  }
+
+  private void populateMetricNames(Set<SystemStreamPartition> ssps) {
+    HashMap<String, String> tags = new HashMap<>();
+    tags.put("client-id", clientId);// this is required by the KafkaConsumer to get the metrics
+
+    for (SystemStreamPartition ssp : ssps) {
+      TopicPartition tp = KafkaSystemConsumer.toTopicPartition(ssp);
+      perPartitionMetrics.put(ssp, new MetricName(tp + ".records-lag", "consumer-fetch-manager-metrics", "", tags));
+    }
+  }
+
+  // The only way to figure out lag for the KafkaConsumer is to look at the metrics after each poll() call.
+  // One of the metrics (records-lag) shows how far behind the HighWatermark the consumer is.
+  // This method populates the lag information for each SSP into latestLags member variable.
+  private void populateCurrentLags(Set<SystemStreamPartition> ssps) {
+
+    Map<MetricName, ? extends Metric> consumerMetrics = kafkaConsumer.metrics();
+
+    // populate the MetricNames first time
+    if (perPartitionMetrics.isEmpty()) {
+      populateMetricNames(ssps);
+    }
+
+    for (SystemStreamPartition ssp : ssps) {
+      MetricName mn = perPartitionMetrics.get(ssp);
+      Metric currentLagM = consumerMetrics.get(mn);
+
+      // High watermark is fixed to be the offset of last available message,
+      // so the lag is now at least 0, which is the same as Samza's definition.
+      // If the lag is not 0, then isAtHead is not true, and kafkaClient keeps polling.
+      long currentLag = (currentLagM != null) ? (long) currentLagM.value() : -1L;
+      latestLags.put(ssp, currentLag);
+
+      // calls the setIsAtHead for the BlockingEnvelopeMap
+      sink.setIsAtHighWatermark(ssp, currentLag == 0);
+    }
+  }
+
+  // Get the latest lag for a specific SSP.
+  private long getLatestLag(SystemStreamPartition ssp) {
+    Long lag = latestLags.get(ssp);
+    if (lag == null) {
+      throw new SamzaException("Unknown/unregistered ssp in latestLags request: " + ssp);
+    }
+    return lag;
+  }
+
+  // Using the consumer to poll the messages from the stream.
```

Comment:
This is a similar comment to pollConsumer. I'd prefer to clarify how they're related. Also, since pollConsumer is only called within this method, this method should be above pollConsumer in this class.

Reply:
I'd propose the following method order for readability:
```
package private Methods:
addTopicPartition
start
isRunning
stop
getFailureCause

private methods:
createProxyThreadRunnable
initializeLags
fetchMessages
pollConsumer
processResults
moveMessagesToTheirQueue
populateCurrentLags (inline populateMetricNames)
updateMetrics (inline getLatestLag)
getRecordSize
```